presto-main/src/main/java/io/prestosql/operator/TopNOperator.java

```diff
@@ -109,72 +108,103 @@ public WorkProcessorOperator create(
+                    n,
+                    sortChannels,
+                    sortOrders,
+                    Optional.of(sourcePageBuffer));
         }
     }
 
-    private final OperatorContext operatorContext;
     private final TopNProcessor topNProcessor;
-    private boolean finishing;
+    private final WorkProcessor<Page> pages;
 
     public TopNOperator(
-            OperatorContext operatorContext,
+            MemoryTrackingContext memoryTrackingContext,
+            WorkProcessor<Page> sourcePages,
             List<Type> types,
             int n,
             List<Integer> sortChannels,
-            List<SortOrder> sortOrders)
+            List<SortOrder> sortOrders,
+            Optional<PageBuffer> pageBuffer)
     {
-        this.operatorContext = operatorContext;
         this.topNProcessor = new TopNProcessor(
-                requireNonNull(operatorContext, "operatorContext is null").aggregateUserMemoryContext(),
+                requireNonNull(memoryTrackingContext, "memoryTrackingContext is null").aggregateUserMemoryContext(),
                 types,
                 n,
                 sortChannels,
                 sortOrders);
 
         if (n == 0) {
-            finishing = true;
+            pages = WorkProcessor.of();
+        }
+        else {
+            pages = sourcePages.transform(new TopNPages());
         }
-    }
 
-    @Override
-    public OperatorContext getOperatorContext()
-    {
-        return operatorContext;
+        pageBuffer.ifPresent(buffer -> buffer.setAddPageListener(() -> {
+            addPage(buffer.poll());
+        }));
     }
 
     @Override
-    public void finish()
+    public WorkProcessor<Page> getOutputPages()
     {
-        finishing = true;
+        return pages;
     }
 
     @Override
-    public boolean isFinished()
-    {
-        return finishing && topNProcessor.noMoreOutput();
-    }
+    public void close()
+            throws Exception
+    {}
 
-    @Override
-    public boolean needsInput()
+    private void addPage(Page page)
     {
-        return !finishing && !topNProcessor.noMoreOutput();
-    }
-
-    @Override
-    public void addInput(Page page)
-    {
-        checkState(!finishing, "Operator is already finishing");
         topNProcessor.addInput(page);
     }
 
-    @Override
-    public Page getOutput()
+    private class TopNPages
+            implements WorkProcessor.Transformation<Page, Page>
     {
-        if (!finishing || topNProcessor.noMoreOutput()) {
-            return null;
+        @Override
+        public TransformationState<Page> process(Page inputPage)
+        {
+            if (inputPage != null) {
+                addPage(inputPage);
+                return TransformationState.needsMoreData();
+            }
+
+            // no more input, return results
+            Page page = null;
+            while (page == null && !topNProcessor.noMoreOutput()) {
+                page = topNProcessor.getOutput();
```

Comment:
does this need to consider yielding?

Reply:
It will quit loop at second iteration at most, see:
`io.prestosql.operator.TopNProcessor#getOutput`